### DISCUSSION TITLE
Feature suggestions (detailed list)

### DISCUSSION THREAD

**Discussion by delidolu1adam** [stoatchat/discussions#828] [Feature Suggestions]
**Customizable usernames**
Optionally customizable usernames. For example: revolt - anyone who types this can add revolt as a friend. No need for the #0001 numerical numbers at the end.

**Phone number verification**
Phone number verification other than email to prevent scammers and spambots. However, you should block toll-free phone numbers in some countries.

**SMS recovery for hacked accounts**
For example, you can easily recover a hacked email and revolt account with a phone number. All you have to do is press the Recover my account button, enter your phone number, and reconnect to your account with the "single-use" password by typing it in the box in the SMS sent to you.

**Desktop and game screen sharing**
You should definitely add Desktop and game screen sharing, one of the most popular features of Discord. This is especially important for helping each other in gaming community groups.

**Shortcut keys**
Settings, friends, mute/enable microphone, mute/enable speaker

**User Profile Badges**
- Loyalty Badge - (1), (2) etc. icons. Shows the date of first membership when hovered over.
- Early Supporter Badge - When hovered over, it can write a nice message like "Our most loyal friend." You can give this badge to everyone who signed up during the development phase.
- Developer Badge - When hovered over, it can write a message like "A devoted person who worked hard to develop Revolt." You can only give this badge to those who contributed to development.
- Donor Enthusiast - When hovered over, it can write a message like "Our most supportive friend." You can only give this badge to those who donated.
- Feedback Enthusiast - You can send this badge to someone who has given feedback at least 10 times.

**Invite to server shortcut in user menu**
Shortcut menu item to invite a user to the server by right-clicking on them.

**User preview in client**
You can make the "Profile picture, Username, status, microphone enable/disable, speaker enable/disable, settings" buttons in the bottom left corner of the client more presentable and useful. Example: https://ibb.co/FgLwr5G
We can also add a "WiFi" icon to a suitable place and show our connection quality as 1-2-3-4-5 bars. When you hover over the WiFi icon, you can also see our PING value.

**Quick channel edit button**
A gear icon on the right side of each channel. Clicking it opens the channel settings and is only visible to those with channel editing permissions. Example: https://ibb.co/mFwZqqF

**Quick invite link generator button**
A plus icon on the right side of each channel. When clicked, it generates an invite link for that channel and copies it to the clipboard. This icon is also only visible to those with invite permissions.

**Slow mode**
Sets the duration of sending messages to the channel. It can be selected from the menu, with a minimum of 5 seconds and a maximum of 7 days.

**Age restricted channel**
Users will be required to verify that they are over the legal age to view channel content.

**Channel password**
Optionally, a password can be assigned to channels, if left blank, the channel can be connected without a password. The channel password can be skipped with a perm as "ignore channel passwords in authorization assignments".

**Creating a survey**
In text channels, the "Upload a file and Create a survey" options can be added as a dropdown menu to the + icon used to upload files. Thus, those who have "Can create a survey" permission in that channel can create a survey.

**Duplicate channel**
We can duplicate a channel with previously set permissions by right-clicking on any text or audio channel.

**BBCODE feature for text channels**
Before sending a message to the text channel, we can select the desired areas of our message and edit them as "bold, italic, strikethrough, quote, code, spoiler". When we select the text, a small popup menu opens and we can click on the buttons from here to perform the action.

**Tagging @everyone , @here and @role**
Allowing those with permissions to tag "@everyone, @here, and @role" in text channels is a required feature, typically used in "Events and announcements" in large community groups.

**Report a message**
The ability to report messages sent to any text channel. There are 2 options to report this message. When you click on the Report icon, a dropdown menu will open. Report or Report to Revolt just below. This will display the normal Reported messages in the server settings under the Reports tab for server administrators. If the messages reported to Revolt are related to "illegal and objectionable activities", the account owner will be sanctioned. If the account owner continues to commit these violations after this sanction, their account will be permanently banned. This can definitely be a very useful and appreciated feature in many areas such as drugs, smuggling, gambling, child abuse.

**Notification message location for client**
The current notification location is at the top and looks very ugly. You can make it more noticeable and more pleasant by making it at the bottom of the window with a red background and white text. :)

**Pinned messages**
The ability to pin messages in text channels can be added. This way, we can pin important messages and view pinned messages in a pop-up window by clicking on the icon in the top right corner of that channel.

**Emoji filter**
You can make emojis more useful and easier to find by separating them with "tab tabs" in the emoji window.

**System message channels**
In addition to the User joined, User left, User kicked, and User banned notification channel assignments, the following can also be added.
Rules, Announcements, System Update Notes

**Hide channels**
If a channel is specified for Rules and Announcements in the server settings, users can optionally hide or re-vise channels other than these.

**Additional features for user profile preview**
1: We can add a note to the user that only we can see. 
2: We can add a nickname to the user. This way, we can see which friend they are even if they change their name.
3: We can see the roles he has on his profile.

**Reset sound and notification settings**
To reset Sound settings and Notification settings to default, a button and explanation text can be added to the bottom of existing pages. This way, even if we change the settings on the page where the sound settings or notification settings are, we can easily return them to default settings.

**Join voice channel with double click**
We need to be able to join any voice channel by double clicking on it, it is very unusual to click on the icon in the upper right corner to join the voice channel. You can also add an icon to the menu in the lower left corner to leave the channel, so that when we click on it, our connection to the channel will be disconnected. :)

**Show users who joined the channel**
List users connected to your voice channel with avatar and display name under the channel. Example: https://ibb.co/NNWcJ24

**Icons for user microphone and speaker status**
Add an icon so that other users can see when the microphone and speaker status of users joining the channel is disabled. Example: https://ibb.co/mFwZqqF

**In addition to system message channels**
Assign default channel to the following channels as an alternative to the "User joined, User left, User kicked, User banned" channel settings in the server settings.
1: Announcements
2: Rules
3: Client updates (optional)
4: AFK (min 5min - max 1 hour)

**Custom server invite link**
A permanent and unique invite link for the server. Anyone with this link can join the server.
Example: https://revolt.chat/kafadartayfa

**Restriction for new members joining the server**
It can be enabled or disabled from the server settings. When enabled; we can determine who can send messages to the server and under what conditions. Example options are listed below.
1: Users who have registered to Revolt for at least 24 hours.
2: Users who have completed email verification.
3: Users who have completed email and phone verification.
4: Users who have joined the server for at least xx minutes.

**A few features for user privacy**
1: Who can add me as a friend? Options: Everyone, friends of friends, no one
2: Who can see my profile details? Options: Everyone, my friends, friends of friends, only me
3: Who can send me private messages? Options: Everyone, my friends, friends of friends, no one

**Role assignment shortcut menu**
We should be able to quickly assign a role to any user in the user list on the server by right-clicking on them and clicking on the Roles menu.

### **I'll update my post as more comes to mind. That's all for now. <3**

**Comment by Oidaho** [comment]
Overall, this is all that is expected of Revolt.
 However, let's remember that this is an open-source solution that students are working on (I understand them perfectly, I'm just like them).
 Therefore, you should not expect all this in the near future.

**Reply by delidolu1adam** [reply]
Hey man, I know this is open source and in development. You don't need to say that because everyone knows it's on Github.com :) I hope you also noticed that the posts users submit under the "Feature Suggestions" category are for suggestions and ideas only. :)

**Comment by Damariobros** [comment]
Perhaps offer a more private option for people who don't want to give their phone number? Also the phone number can be changed by a would-be hacker. Unless you mean for Stoat to retain the phone number on file for a time after a phone number is changed?

For account recovery, allow users to set trusted contacts who can vouch for them. In a support ticket, trusted contacts could be able to vouch that this is not the account's normal behavior and that they've been contacted by the account owner through another channel of communication.

Or perhaps, allow for some kind of recovery key that would, like, invalidate all sessions, revert all account information/security changes by 7 days (i.e. undo an email change, undo 2fa changes, undo username/handle change, restore pfp and bio and display name, etc.), lock the account temporarily by invalidating the current password, and send a password reset email?

Maybe to account for compromised recovery keys, as well as for compromised accounts having keys rotated, such a key could be combined with a PIN. If the key is compromised, you rotate the key, attacker doesn't know the PIN. If the account is compromised and the key is rotated, enter the old key with the PIN within 7 days and it'll be valid even though it's been rotated.

Maybe advanced users could add an OpenPGP public key to their account file which they could use to help affirm their identity to customer support.

Maybe Stoat could offer proactive measures for limiting the damage account compromises can cause. An optional account PIN could be set like the kind Roblox used to have for their parental controls system, where account settings such as personal details, password, and 2fa settings are frozen until you supply the PIN, which thaws the settings for 5 minutes. A PIN could be reset via email code.

(Pardon the technical reads, I enjoy thinking things through and creating detailed, granular suggestions.)
